JokeDAO/JokeDaoV2Dev
# feat: separate tokens and requirements for submissions and voting

# Description
[Clarity task: GEN-44](https://app.clarity.so/jokedaosbase3/group/GEN/projects)
> Separate tokens and requirements for submissions and voting

### Contest creation Wizard form
 - Allow users to specify a different token to use for proposals
 - Add "Mint proposal token" dialog to contest creation form
 - Add an indication to use a  token that implements the `IERC20` interface for voting/submission tokens
![Screenshot_20220905_164859](https://user-images.githubusercontent.com/15010369/188484266-5479d1bf-f33e-42bf-9039-94192fbb8cae.png)
 
### Contest page
#### `useContest` hook
- Add `currentUserSubmitProposalTokensAmount` to the `store` ; this value will hold the current user's current amount of the token required to submit a proposal
- Add `checkCurrentUserAmountOfProposalTokens()` function to the hook ; this function calls `getCurrentSubmissionTokenVotes` and store the returned value in ` currentUserSubmitProposalTokensAmount`
- Add `submitProposalToken`, `submitProposalTokenAddress` and their update function to the store ; these values will hold the submission proposal tokens info (for legacy contests, it will store the voting token info)

#### `Sidebar`,  `ListProposals`,  `DialogModalSendProposal`
- In the "submit proposal" button, instead of using `currentUserAvailableVotesAmount` value to enable/disable the button, use `currentUserSubmitProposalTokensAmount` value instead 
 
### Rules page
- Separate
![Screenshot 2022-09-05 at 17-36-19 JokeDAO 🃏 An open-source collaborative decision-making platform](https://user-images.githubusercontent.com/15010369/188484413-20d4b365-1f55-4f2c-9b86-cc8883da33c9.png)
 the "Token" section in  two sections, "Voting token" and "Submit proposals token"


## Dialogs
- Add error messages to all Dialogs (use the til now unused `textError` prop)

## Type of change
- [x] New feature with breaking changes

# How Has This Been Tested
## Contest creation
1. Go to create contest page
2. Create a new token then click on "next" to fo to "set rules" step
![Screenshot 2022-08-22 at 16-27-28 JokeDAO 🃏 An open-source collaborative decision-making platform](https://user-images.githubusercontent.com/15010369/185947210-93c09412-9cae-4c1f-9a39-d75a4ab12226.png)

3. In the form, in the submissions section, you should  see a new fieldset with radio buttons: 
   - Option 1 should display "Use the same token for both submitting proposals and voting". This option should be selected by default
   - Option 2 should display "Require separate submission token" + a text input that accepts an ethereum address + a "Mint new token" button
 
![Screenshot 2022-08-22 at 16-28-05 JokeDAO 🃏 An open-source collaborative decision-making platform](https://user-images.githubusercontent.com/15010369/185947259-cbec530a-4ebb-4d4b-be5d-db0ac86b5997.png)


4. Click on "Mint new token"
5. A dialog modal should appear with the create token form
![Screenshot 2022-08-22 at 16-28-42 JokeDAO 🃏 An open-source collaborative decision-making platform](https://user-images.githubusercontent.com/15010369/185947280-b7226c27-6ddc-4504-95c5-ef6ae9a2e675.png)

6. Fill in the fields
![Screenshot 2022-08-22 at 16-28-42 JokeDAO 🃏 An open-source collaborative decision-making platform](https://user-images.githubusercontent.com/15010369/185947383-034f1318-0ef9-4dac-a23f-10ca0784022b.png)


7. Click on "Mint"
8. The stepper should appear
![Screenshot 2022-08-22 at 16-28-55 JokeDAO 🃏 An open-source collaborative decision-making platform](https://user-images.githubusercontent.com/15010369/185947413-818d5ad8-a0a7-4c93-a1d9-d4b11b12e3bd.png)

9. Approve the transaction in your wallet
10. When the transaction is successful, the form should be removed, and a button that says "Go back to contest creation" should appear
![Screenshot 2022-08-22 at 16-29-16 JokeDAO 🃏 An open-source collaborative decision-making platform](https://user-images.githubusercontent.com/15010369/185947471-49bd0337-b0c2-4c99-8baf-d404edaac304.png)

11. If the transaction failed, you should still see the form, and a "try again" button should appear. 

12. Back in the contest creation form ; "Require separate submission token" input should have the newly minted token address as its value
![Screenshot 2022-08-22 at 16-29-24 JokeDAO 🃏 An open-source collaborative decision-making platform](https://user-images.githubusercontent.com/15010369/185947504-67a09897-9f74-490f-a5d9-4a9e7cffbb06.png)

13. Add the rest of the rules 
14. Click on "Create contest". The modal with the steps should appear.
15. Sign the transaction.
16.  If the transaction is successful, the link to the newlt created contest should appear
17. Go to the newly created contest
18. The contest and its data should be displayed

For the following cases, switch to use an account that has no "proposal" token
### Case 1: no minimum amount of token required to submit a proposal
1. The "Submit proposal" button should be visible and enabled

### Case 2: a minimum amount of token is required to submit a proposal
1. The "Submit proposal" button should be visible and but disabled

For the following cases, switch to use an account that has proposal tokens
### Case 3: a minimum amount of token is required to submit a proposal, but the current wallet doesn't have enough tokens
1. The "Submit proposal" button should be visible and but disabled
![Screenshot_20220905_153902](https://user-images.githubusercontent.com/15010369/188484620-a356b878-83ff-4f39-9fcc-420a73e1676c.png)


### Case 4: a minimum amount of token is required to submit a proposal, the current wallet has enough tokens
1. The "Submit proposal" button should be visible and enabled
![Screenshot_20220905_153759](https://user-images.githubusercontent.com/15010369/188484561-0c87094e-efa4-4a77-a834-76d825ed0275.png)


## Legacy contests
1. Legacy contests that an ABI that doesn't include the newest functions should display properly (checked contest: `/polygon/0x3aCD536121141BDA3E8C0910E21802DA5Be3fCae` 

---

## Checklist
- [x] I can mint a token
- [x] I can create a contest
- [x] I can  display previous/upcoming/live contests index and my newly created contest is there
- [x] I can display a newly created contest page
- [x] I can display a legacy contest page (a contest which ABI doesn't include the new feature implemented here)
